File: be/models/notification.model.js
const db = require("../configs/db");

const notificationSchema = new db.mongoose.Schema(
  {
    user_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "userModel",
      required: true,
    },
    title: { type: String, required: true },
    body: { type: String, required: true },
    type: {
      type: String,
      required: true,
      enum: [
        "order",
        "voucher",
        "system",
        "PAYMENT_SUCCESS",
        "PAYMENT_FAILED",
        "ORDER_CREATED",
        "ORDER_CONFIRMED",
        "ORDER_PREPARING",
        "ORDER_SHIPPING",
        "ORDER_COMPLETED",
        "ORDER_CANCELLED"
      ],
    },
    data: { type: Object, default: null },
    is_read: { type: Boolean, required: true, default: false },
    sent_at: { type: Date, default: null },
    // Ai tạo ra thông báo này — để lịch sử "Thông báo đã gửi" ở Admin/Shop
    // Portal lọc đúng phạm vi (system = tự động sinh, không phải ai chủ động gửi).
    sender_type: {
      type: String,
      enum: ["admin", "shop", "system"],
      default: "system",
    },
    sender_shop_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "shopModel",
      default: null,
    },
    deleted_at: { type: Date, default: null },
  },
  { collection: "notifications", timestamps: true },
);

let notificationModel = db.mongoose.model(
  "notificationModel",
  notificationSchema,
);
module.exports = { notificationModel };

File: tuhubread/lib/models/notification.model.dart
class NotificationModel {
  final String id;
  final String title;
  final String message;
  final String type;
  final bool isRead;
  final DateTime sentAt;
  final String? orderId;

  NotificationModel({
    required this.id,
    required this.title,
    required this.message,
    required this.type,
    required this.isRead,
    required this.sentAt,
    this.orderId,
  });

  factory NotificationModel.fromJson(Map<String, dynamic> json) {
    final data = json['data'] as Map<String, dynamic>?;
    return NotificationModel(
      id: json['_id'] as String,
      title: json['title'] as String? ?? '',
      message: json['body'] as String? ?? json['message'] as String? ?? '',
      type: json['type'] as String? ?? 'system',
      isRead: json['is_read'] as bool? ?? false,
      sentAt: json['sent_at'] != null 
          ? DateTime.parse(json['sent_at'] as String)
          : (json['createdAt'] != null ? DateTime.parse(json['createdAt'] as String) : DateTime.now()),
      orderId: data?['orderId'] as String?,
    );
  }
}
